tdrs-backend/tdpservice/search_indexes/parsers/tanf_parser.py

```diff
@@ -0,0 +1,269 @@
+                        line_number, field.start-1, field.name, field.length, content)
+            content_is_valid = False
+            continue
+        # check if content is type string or integer
+        if field.type == 'Numeric':
+            try:
+                content = int(content)
+            except ValueError:
+                logger.warn('[LineNo:%d, col%d] Expected field "%s" to be numeric, got: "%s"',
+                            line_number, field.start-1, field.name, content)
+                content_is_valid = False
+                continue
+        elif field.type == 'Alphanumeric':
+            pass  # maybe we can regex check some of these later
+        # The below is extremely spammy, turn on selectively.
+        # logger.debug('field: %s\t::content: "%s"\t::end: %s', field.name, content, field.end)
+
+        if content_is_valid:
+            setattr(t1, field.name, content)
+
+    if content_is_valid is False:
+        logger.warn('Content is not valid, skipping model creation.')
+        return
+
+        '''
+        '''
+
+    # try:
+    # t1.full_clean()
+    t1.save()
+    '''
+    # holdovers for 1354
+        ParserLog.objects.create(
+            data_file=datafile,
+            status=ParserLog.Status.ACCEPTED,
+        )
+
+    except ValidationError as e:
+        return ParserLog.objects.create(
+            data_file=datafile,
+            status=ParserLog.Status.ACCEPTED_WITH_ERRORS,
+            errors=e.message
+        )
+    '''
+
+# TODO: def closed_case_data(datafile):
+
+# TODO: def aggregate_data(datafile):
+
+# TODO: def stratum_data(datafile):
+
+def parse(datafile):
+    """Parse the datafile into the search_index model."""
+    logger.info('Parsing TANF datafile: %s', datafile)
+
+    datafile.seek(0)  # ensure we are at the beginning of the file
+    line_number = 0
+    for raw_line in datafile:
+        line_number += 1
+        # logger.debug('Parsing this line: "%s"', raw_line)
+        if isinstance(raw_line, bytes):
+            # logger.info("Line is bytes, decoding to string...")
+            raw_line = raw_line.decode()
+        line = raw_line.strip('\r\n')
+
+        record_type = get_record_type(line)
+        # logger.debug('Parsing as type %s this line: "%s"', record_type, line)
+
+        if record_type == 'HE' or record_type == 'TR':
+            # Header/trailers do not differ between types, this is part of preparsing.
+            continue
+        elif record_type == 'T1':
+            expected_line_length = 156  # we will need to adjust for other types
+            actual_line_length = len(line)
+            if actual_line_length != expected_line_length:
+                logger.error('Expected line length of 156, got: %s', actual_line_length)
+                # should be added to parser log in #1354
+                continue
+            else:
+                active_t1(line, line_number)
```

Comment:
The nested functions go pretty deep between parse and active_t1. I am wondering if we can write a custom iterator or zip function that will look cleaner and be easier to extend. But I think this is a great start.

Reply:
I'm not clear on what you're proposing. You want an iterator for what, the lines of the file to replace this for-loop?